fix: Preserve order of jest calls when hoisting

tgriesser opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

It looks like the changes to hoisting in #9806 have incorrectly ordered hoisted `jest` in the same block, because they are unshifted on the containing scope individually, rather than collectively after walking the AST.

## Test plan

Added an integration test in `e2e/babel-plugin-jest-hoist`. 
```js
import a from '../__test_modules__/a';

// These will all be hoisted above imports
jest.enableAutomock();
jest.disableAutomock();
```

Before the change:

```ts
'use strict';

// These will all be hoisted above imports
_getJestObj().disableAutomock();

_getJestObj().enableAutomock();

// ... rest of spec
```

After change:

```ts
'use strict';

// These will all be hoisted above imports
_getJestObj().enableAutomock();

_getJestObj().disableAutomock();

// ... rest of spec
```

SimenB commented:
This change makes sense to me, but CI is very unhappy. I'm assuming this somehow broke our of usage of `jest.mock`?

Also, please add a changelog entry 🙂

tgriesser commented:
Haven't quite figured out how to adapt the Babel plugin here correctly. Tried a few things, including collecting the nodes during traversal and then adding the nodes on exit:

```ts
const onExitBlock = (node: NodePath<Block>) => {
  const hoistNodes = node.getData('__babelJestHoistNodes');
  if (hoistNodes) {
    node.unshiftContainer('body', hoistNodes);
  }
};

// ...
{
  post({path: program}: {path: NodePath<Program>}) {
    program.traverse({
      Block: {
        exit: onExitBlock,
      },
      CallExpression: callExpr => {
        const {
          node: {callee},
        } = callExpr;
        if (
          isIdentifier(callee) &&
          callee.name === this.jestObjGetterIdentifier?.name
        ) {
          const mockStmt = callExpr.getStatementParent();

          if (mockStmt) {
            const mockStmtNode = mockStmt.node;
            const mockStmtParent = mockStmt.parentPath;
            if (mockStmtParent.isBlock()) {
              mockStmt.remove();
              const hoistNodes =
                mockStmtParent.getData('__babelJestHoistNodes') ?? [];
              hoistNodes.push(mockStmtNode);
              mockStmtParent.setData('__babelJestHoistNodes', hoistNodes);
            }
          }
        }
      },
    });
    onExitBlock(program as NodePath<Block>);
  },
}
```

But for some reason the nodes aren't being added properly. Will try a few more things, but would love some eyes from anyone a little more versed in babel transforms.

SimenB commented:
@jeysal @nicolo-ribaudo ideas? 😀 I still suck at AST transformations, so not much help here 😛

We should probably have some more direct unit tests for this like discussed in https://github.com/facebook/jest/pull/9806#issuecomment-620492720

tgriesser commented:
@SimenB ok, tried a new approach which seems to work decent enough - added some unit tests as you mentioned and fixed an e2e test which seemed to only be passing due to the previously incorrect hoisting behavior.

tgriesser commented on e2e/babel-plugin-jest-hoist/__tests__/hoistOrder.test.js:
```diff
@@ -0,0 +1,21 @@
+import a from '../__test_modules__/a';
+
+// These will all be hoisted above imports
+jest.enableAutomock();
```
Can probably remove this e2e test now that there's more direct snapshot tests, will do that after the latest tests complete.

SimenB commented on e2e/babel-plugin-jest-hoist/__tests__/integration.test.js:
```diff
@@ -34,6 +34,10 @@ let e;
   jest.unmock('../__test_modules__/e');
 })();
 
+// Variable names prefixed with `mock` (ignore case) should not throw as out-of-scope
```
looking at https://github.com/facebook/jest/pull/9806/files#r415385512, I think the one you move should just be `g` instead? Not sure, looks like a half completed refactor? Would love @jeysal's comments on this

jeysal replied:
Yeah I agree, we can use a different file and avoid inserting this in the middle of the 'These will all be hoisted above imports' cases

SimenB reviewed (approved):
yay, thanks for adding a start for proper unit tests 👍

tgriesser commented:
Sidenote: In adding the unit tests, and trying it out with some other issues around `@jest/globals` like #10218 and #10147, I feel like there's a number of cases that aren't fully covered by the globals change/refactor. 

Hoping this will be a good starting place for addressing some of those issues.

jeysal reviewed (changes requested):
Thanks for the contribution, I agree that order preservation is a valuable guarantee to give!

Regarding the tests I also think it's important to build up a more maintainable test suite than the current one. I think long term we want to go towards exec tests, because snapshot tests like these will in large numbers become unmaintainable as well, because different but equally valid output in 20 snapshots is hard to review. However, this is a good start for sure.

Regarding the approach to fixing this in the Babel plugin using stateful logic, I'm not so sure. Considering that we may have a couple more edge cases to fix in the future as mentioned, I'm worried that the state of the 'tags' on nodes may become confusing. Have you considered the following rough structure, which unless I'm missing something should allow working statelessly close to the existing code:
```
program.traverse
  BlockStatement|Program
    let firstStatementOfBlock = ...
    block.traverse
      CallExpression
        (The existing code except mockStmtNode is inserted before firstStatementOfBlock)
```

SimenB commented:
@tgriesser thoughts on @jeysal's suggestion?

tgriesser commented:
Haven't had a chance to come back to this. If one of y'all want to take a stab at it, feel free.

SimenB commented:
That's perfectly fine @tgriesser, thanks for getting this started!

@jeysal would you be up for finishing this? If not I'll just land as is as it's a bug fix

jeysal commented:
Yep, I'll give it a shot, I'll have more time again from Nov I think

jeysal commented:
Pushed a stateless solution that passes the added tests, just noticed it seems to fail one e2e test though, will TAL

jeysal commented:
Nvm that was actually a mistake with the minor comment about moving things around in the test. The code appears to work fine

jeysal reviewed (approved):
LGTM now obviously lol

jeysal commented:
@tgriesser feel free to leave a review on my refactor if you want, or if you had an original problem reproducing you could double check that this fixes it :upside_down_face:

tgriesser commented:
@jeysal Looks great! One note would be that `blacklist` was [renamed as `denylist`](https://github.com/babel/babel/pull/11791) in babel 7.11 and will be fully removed in babel 8. I assume based on the currently dependency ranges it'd be best to keep as `blacklist` for now though, unless you wanted to bump deps to newer versions of babel.

SimenB reviewed (approved):
This is great 👍

SimenB commented:
I'll bump traverse version and migrate the name in a separate PR
